File: orchestrator/app/services/runner_service.py
import asyncio
import httpx

async def wait_for_runner(base_url: str) -> bool:
    """Wait for the runner FastAPI server to start responding."""
    url = f"{base_url.rstrip('/')}/docs"
    async with httpx.AsyncClient() as client:
        for _ in range(30): 
            try:
                resp = await client.get(url, timeout=1.0)
                if resp.status_code == 200:
                    return True
            except httpx.RequestError:
                pass
            await asyncio.sleep(0.5)
    return False

async def trigger_runner_start(base_url: str, project_id: str) -> None:
    """Tell the runner to initialize the mounted project workspace."""
    print(f"[RunnerService] Triggering runner workspace init for projectId={project_id} via {base_url}")
    async with httpx.AsyncClient() as client:
        try:
            resp = await client.post(
                f"{base_url.rstrip('/')}/start",
                json={"projectId": project_id},
                timeout=30.0
            )
            if resp.status_code != 200:
                print(f"[RunnerService] Warning: Runner start returned status {resp.status_code}: {resp.text}")
        except Exception as e:
            print(f"[RunnerService] Error triggering runner start: {e}")

File: orchestrator/src/services/runner-service.ts
import axios from "axios";

export async function waitForRunner(baseUrl: string): Promise<boolean> {
  const url = `${baseUrl.replace(/\/$/, "")}/docs`;

  for (let i = 0; i < 30; i++) {
    try {
      const response = await axios.get(url, { timeout: 1000 });
      if (response.status === 200) {
        return true;
      }
    } catch (error) {
      // Ignore request error and retry
    }
    // Sleep for 500ms
    await new Promise((resolve) => setTimeout(resolve, 500));
  }

  return false;
}

export async function triggerRunnerStart(baseUrl: string, projectId: string, projectType: string): Promise<void> {
  const url = `${baseUrl.replace(/\/$/, "")}/start`;
  console.log(`[RunnerService] Triggering runner workspace init for projectId=${projectId}, projectType=${projectType} via ${baseUrl}`);

  try {
    const response = await axios.post(
      url,
      { projectId, projectType },
      { timeout: 30000 }
    );
    if (response.status !== 200) {
      console.warn(`[RunnerService] Warning: Runner start returned status ${response.status}: ${JSON.stringify(response.data)}`);
    }
  } catch (error: any) {
    console.error(`[RunnerService] Error triggering runner start: ${error.message || error}`);
  }
}
